### PR TITLE
[stdlib] Make use of protocol requirements to convert from concrete floating-point types

### PR DESCRIPTION
**Pull request by xwu**
This PR is about picking the low-hanging fruit in terms of recovering performance during generic floating-point conversion. It is _so_ low-hanging, in fact, it's actually partly a reversion to the original placeholder implementation of generic floating-point conversion.

In brief, we never removed the requirements for `init(_: Float)`, `init(_: Double)`, and `init(_: Float80)` from the protocol, and now that we have ABI stability, these requirements will be present forever. So, we can call them from the generic implementation.

There is one caveat to be highlighted: As in other situations (see #30417), the compiler regards the generic initializer itself as a suitable default implementation of `init(_: Float)` and friends, and therefore it will not warn if a concrete type does not provide its own implementations of the latter. Currently, this means that the concrete type will unwittingly have a very slow implementation; with this change, the concrete type will have an implementation that crashes at runtime due to infinite recursion. I think it's arguable as to which one is the worse outcome; this PR doesn't cause or otherwise change the underlying issue.

### CODE REVIEW

**Comment by xwu** [comment]
cc @troughton

**Comment by stephentyrone** [comment]
@swift-ci test performance.

**Comment by stephentyrone** [comment]
@swift-ci benchmark

File: stdlib/public/core/FloatingPoint.swift [hunk 1]
```diff
@@ -1890,7 +1890,18 @@ extension BinaryFloatingPoint {
   /// - Parameter value: A floating-point value to be converted.
   @inlinable
   public init<Source: BinaryFloatingPoint>(_ value: Source) {
-    self = Self._convert(from: value).value
+    switch value {
+    case let value_ as Float:
```
Comment (stephentyrone) [inline stdlib/public/core/FloatingPoint.swift:1894]: This should handle `Float16` as well (implemented as `self = Self(Float(value_))` in that case).

**Comment by xwu** [comment]
There's a benchmark PR not yet merged, by the way.

**Comment by xwu** [comment]
@swift-ci test

**Comment by xwu** [comment]
@swift-ci test

**Comment by xwu** [comment]
@swift-ci test

File: stdlib/public/core/FloatingPoint.swift (hunk 1 above)
Comment (xwu) [reply stdlib/public/core/FloatingPoint.swift:1902]: @stephentyrone The required checks for the availability of `Float16` are...not pretty. See the commit "🤮" for details.

**Comment by xwu** [comment]
@swift-ci benchmark

**Comment by xwu** [comment]
@swift-ci test Linux platform

**Comment by xwu** [comment]
@swift-ci benchmark

**Comment by xwu** [comment]
@swift-ci please benchmark

**Comment by xwu** [comment]
Hmm

@swift-ci please benchmark

**Pull request merged**